### ISSUE TITLE
[Task]: Add tests to Kafka SDF and fix known and discovered issues

### ISSUE THREAD

**Issue by johnjcasey** [apache/beam#22303]
### What needs to happen?

Kafka IO SDF currently has an issue where it is not restarting properly: https://github.com/apache/beam/issues/21730

In addition, there was a separate issue where it would not commit back to Kafka correctly. 

As such, it appears that Kafka SDF has not been sufficiently tested, and the scope of this task is to test it more robustly, and fix both the linked issue, and any other issues discovered.

### Issue Priority

Priority: 1

### Issue Component

Component: io-java-kafka

**Comment by johnjcasey** [comment]
@Abacn this issue can be used to track the kafka testing improvements

**Comment by johnjcasey** [comment]
https://github.com/apache/beam/issues/22715 also

**Comment by pabloem** [comment]
how should we triage this? should we assign it to @johnjcasey or @Abacn ?

**Comment by Abacn** [comment]
Apparently Kafka performance test is failing again after #24879 in. This is because we removed `shuffle=appliance` there and then read from shuffle becomes faster. There are some reports that runner v2 is putting more pressure on kafka, and resulting in higher error rate. Now kafka server is overwhelmed with write requests and making write pipeline fail.

This reveals real issue of the code.

**Comment by johnjcasey** [comment]
Closing as out of date. SDF Kafka has been significantly reworked, and is now in a much better state
